### PR TITLE
Add `--imports` flag

### PR DESCRIPTION
**Pull request by d3vco**
- Resolves issue #6
- New flag `--imports`
- Bumps `goblin` to v0.4.3
- Uses `[patch]` to pull in the GitHub-hosted `checksec` for compatibility with `goblin` v0.4.3
- Refactors handling of Multi-Arch Mach-O binaries

### CODE REVIEW

**Review by tnballo — CHANGES_REQUESTED** [review]
Apologize for the delayed feedback, this PR is fantastic! 

Thank you for supporting all 3 file formats in a unified way and digging into some really advanced usage of `goblin`. I learned a bit from reviewing this.

Unfortunately, we can't publish this killer feature to crates.io until `checksec` publishes the upgrade to `goblin` v0.4.3+, because `cargo publish` requires all dependencies to be crates.io hosted (e.g. not just GitHub, [even with `[patch]`](https://github.com/rust-lang/cargo/issues/6738)). Just now remembering that's how this works (which makes sense, otherwise crate.io would be reliant on GitHub to not break builds).

In the interim, I'll update the `README.md`'s "Quickstart" to prefer building the CLI by cloning the repo instead of `cargo install`, to get the latest and greatest version with the `--imports` flag!

**Comment by d3vco** [comment]
Thanks for the thorough feedback @tnballo!

Since we can't publish these changes until `checksec` is updated, I will work on the improvements you've suggested and any others that I that come up while I'm going back over the code.

Todos:
- [ ] Use Rust `Default` trait in `Import` struct
- [ ] Pass `usize` by value
- [ ] Better use of `goblin` constants (replacing magic numbers / hex codes)
- [ ] Improve output (alphabetical, consistent formatting, etc)

**Comment by tnballo** [comment]
Sounds great @d3vco , thanks! Looking forward to merging whenever ready and pushing to crates.io ASAP.

Have opened a PR on the `checksec` repo to hopefully unblock us on that front.

File: src/cli/imports.rs [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+
+            Some(offset)
+        }
+
+        fn get_plt_offset(elf: &goblin::elf::Elf, sym_idx: &usize) -> Option<u64> {
+            let reloc_idx = elf.pltrelocs.iter().position(|r| r.r_sym == *sym_idx)?;
+
+            let mut offset = (reloc_idx as u64 + 1) * 16;
+
+            if let Some(plt) = &elf
```
[the remark this answers is not in the document]
Comment (d3vco) [reply src/cli/imports.rs:59]: I wrote it this way assuming there was a possibility that the `.plt` section header could be stripped, preventing the function (with its current logic) from locating the `.plt` base address. The section would still exist, so a user may still find it useful to have the calculated offset as output, in case they can find the base `.plt` address another way.

Does that make sense or is my hypothetical case unrealistic?

File: src/cli/imports.rs [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+                return None;
+            };
+
+            Some(offset)
+        }
+
+        fn get_plt_offset(elf: &goblin::elf::Elf, sym_idx: &usize) -> Option<u64> {
+            let reloc_idx = elf.pltrelocs.iter().position(|r| r.r_sym == *sym_idx)?;
+
+            let mut offset = (reloc_idx as u64 + 1) * 16;
```
[the remark this answers is not in the document]
Comment (d3vco) [reply src/cli/imports.rs:57]: That case should actually be relatively easy to capture with `goblin`'s [`goblin::elf::dynamic::DF_BIND_NOW`](https://docs.rs/goblin/latest/goblin/elf/dynamic/constant.DF_BIND_NOW.html#) constant. I'll test it out.

[the remark this answers is not in the document]
Comment (tnballo) [reply src/cli/imports.rs]: Gotcha, great to know! Thanks for digging into this and double checking.

File: src/cli/imports.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (tnballo) [reply src/cli/imports.rs]: I see! Agree with you then - returning a calculated offset seems preferable to erroring entirely.

Good call on considering stripped binaries. However, I don't *think* the PLT section can be stripped. Not 100% sure, but here's what I learned in attempting to strip it:

* `gcc -s` and `strip --strip-all` both leave a `.plt` section in tact (checking via `readelf --sections <name> | grep plt`), although symbol stripping works as intended. `strip --remove-section=plt` didn't work either.

* Much to my surprise, even compiling a stripped, static binary with `gcc <name>.c -o <name> -s --static -fno-plt` seems to leave `.plt` in there. Per the docs, `-fno-plt` seems [to be an optimization](https://man7.org/linux/man-pages/man1/gcc.1.html) but not a way to actually remove the `.plt` section.

To double check the disassembly (inspired by your other comment), I created a `hello_world` program:

```c
#include <stdio.h>
int main() {
   printf("Hello, ELF World!\n");
   return 0;
}
```

To test the effect of `-fno-plt` in isolation - compiled with `gcc hello.c -o hello`, then ran `gdb -batch -ex "disassemble main" hello`:

```
Dump of assembler code for function main:
   0x0000000000001149 <+0>:     endbr64
   0x000000000000114d <+4>:     push   %rbp
   0x000000000000114e <+5>:     mov    %rsp,%rbp
   0x0000000000001151 <+8>:     lea    0xeac(%rip),%rdi        # 0x2004
   0x0000000000001158 <+15>:    callq  0x1050 <puts@plt>
   0x000000000000115d <+20>:    mov    $0x0,%eax
   0x0000000000001162 <+25>:    pop    %rbp
   0x0000000000001163 <+26>:    retq
End of assembler dump.
```

Doing the same but adding `-fno-plt` to the `gcc` params:

```
Dump of assembler code for function main:
   0x0000000000001129 <+0>:     endbr64
   0x000000000000112d <+4>:     push   %rbp
   0x000000000000112e <+5>:     mov    %rsp,%rbp
   0x0000000000001131 <+8>:     lea    0xecc(%rip),%rdi        # 0x2004
   0x0000000000001138 <+15>:    callq  *0x2e9a(%rip)        # 0x3fd8
   0x000000000000113e <+21>:    mov    $0x0,%eax
   0x0000000000001143 <+26>:    pop    %rbp
   0x0000000000001144 <+27>:    retq
End of assembler dump.
```

So `-fno-plt` does indeed inline call resolution.

In the latter case, the `--imports` output is empty (which is fine, GOT stuff got inlined directly, it would be pretty crazy to have to resolve such an optimization even if technically possible!). BUT there is still a `.plt` section in the binary.

So, in conclusion: I couldn't strip the PLT no matter how hard I tried. Learned that the `-fno-plt` flag doesn't actually mean "no plt", although it does create another edge case (don't worry about it, seriously - the ELF format has endless edge cases).

[the remark this answers is not in the document]
Comment (d3vco) [reply src/cli/imports.rs]: Awesome, thanks for looking into this! I took your advice and updated the `get_plt_address` function to return `None` if the `.plt` lookup fails. I also explicitly ensure we are dealing with a `JUMP_SLOT` relocation to try to cut out a few more of those endless edge cases.

https://github.com/d3vco/xgadget/blob/ac67ffc7ace35f35fb3a7c14056fd1da22427f72/src/cli/imports.rs#L402-L419

**Comment by d3vco** [comment]
@tnballo , I have completed all of my todos, so I will (try to) stop making new changes at this point. Happy to update / optimize anything you find in your review.

File: src/cli/imports.rs [hunk 3]
```diff
@@ -213,33 +141,337 @@ impl fmt::Display for Import {
...
+    }
+
+    // collect PLT relocations
+    let mut plt_imports = Vec::new();
+    for reloc in elf.pltrelocs.iter() {
+        if let Some(sym) = elf.dynsyms.get(reloc.r_sym) {
+            let print_vec = Import::from_elf(elf, &sym, &reloc, no_color).get_print_vec();
+            // update column width to match longest string in column
+            for (idx, print_str) in print_vec.iter().enumerate() {
+                col_width[idx] = cmp::max(col_width[idx], print_str.len());
```
Comment (tnballo) [inline src/cli/imports.rs:196]: Nice handling of column width, like the new table-like formatting!

File: src/cli/imports.rs [hunk 4]
```diff
@@ -213,33 +141,337 @@ impl fmt::Display for Import {
...
+
+        imports.push(print_vec);
+    }
+
+    // print collected imports
+    println!("Imports:");
+    print_imports(&headers, imports, &col_width, no_color);
+}
+
+fn print_imports(
```
Comment (tnballo) [inline src/cli/imports.rs:306]: Good design for each file format's parsing function to collect/manipulate the same data and have a common printer function.

File: src/cli/imports.rs [hunk 5]
```diff
@@ -213,33 +141,337 @@ impl fmt::Display for Import {
...
+            })?
+            .sh_addr;
+
+        return Some(plt_base + offset);
+    }
+
+    None
+}
+
+fn get_symbol_r_type_64(sym_type: u32) -> String {
```
Comment (tnballo) [inline src/cli/imports.rs:391]: Thanks for adding these lookup tables!

Was wondering if this sort of pretty printing should be a feature of `goblin` itself, but these are actually constants that already implement [`Display`](https://doc.rust-lang.org/std/fmt/trait.Display.html) to print the number itself (not helpful for us).

**Review by tnballo — APPROVED** [review]
Awesome - thanks for designing, implementing, and iterating this feature!

All tests are passing and the `--import` flag output looks great.

`checksec` has updated its crates.io version, so will similarly publish this now!

**Comment by tnballo** [comment]
It's now pushed to crates.io :tada: 

Thanks again, I think the `--imports` flag will be handy for many people!
